### PR TITLE
Post Revisions: Use diffs from the API instead of diffing locally on the front end

### LINKED ISSUE #20583: Editor: console warning/error `openPostRevisionsDialog` when loading editor with history
#### Steps to reproduce
1. Starting at URL: calypso.localhost:3000/posts
2. Edit a post with revision history

#### What I expected

No consoler errors/warnings

#### What happened instead

```
index.jsx:97 Warning: Failed prop type: The prop `openPostRevisionsDialog` is marked as required in `HistoryButton`, but its value is `undefined`.
    in HistoryButton (created by Localized(HistoryButton))
    in Localized(HistoryButton) (created by Connect(Localized(HistoryButton)))
    in Connect(Localized(HistoryButton)) (created by EditorGroundControl)
    in div (created by EditorGroundControl)
    in div (created by Card)
    in Card (created by EditorGroundControl)
    in EditorGroundControl (created by Localized(EditorGroundControl))
    in Localized(EditorGroundControl) (created by PostEditor)
    in div (created by PostEditor)
    in div (created by PostEditor)
    in PostEditor (created by Localized(PostEditor))
    in Localized(PostEditor) (created by ProtectedFormComponent)
    in ProtectedFormComponent (created by Connect(ProtectedFormComponent))
    in Connect(ProtectedFormComponent)
    in Provider

```
#### Browser / OS version

Chrome macOS

#### Screenshot / Video

screen shot 2017-12-07 at 12 02 11 pm

#### Context / Source
Found testing https://github.com/Automattic/wp-calypso/pull/20326

### PR DESCRIPTION
**Pull request by jblz**
The current implementation computes the diff between revisions in the browser / device.

This uses the diffs computed by the REST API server instead. As a result, it is enabled on mobile devices now and performs a lot better than the client-side diffing.

fixes #20583 as well

### Manual testing

* ~~Apply API patch in D8615~~ endpoint is live!
* Run this branch
* Edit a post with revisions
* Click the History button
  * Diffs should be shown and UI fully functional
  * Clicking `Load` should put the "compare from" revision's title & content into the Editor
* Repeat with the `?flags=-post-editor/revisions` query arg
  * The link to wp-admin revisions should be visible 
* Test on Mobile
  * It will show the `History` button in landscape orientation & function well.
  * Rotating back to portrait mode will still be functional, but with the list of revisions scrolling horizontally at the top of the modal

### Automated testing

`npm run test-client:watch client/state/posts/revisions client/state/selectors/test/*revision*`

### CODE REVIEW

File: client/state/posts/revisions/reducer.js [hunk 1]
```diff
@@ -20,7 +20,56 @@ import {
...
+const isNonNegativeInteger = t => isInteger( t ) && t >= 0;
+
+export const diffs = createReducer(
+	{},
+	{
+		[ POST_REVISIONS_RECEIVE ]: (
+			state,
+			{ diffs: diffsFromServer, postId, revisions, siteId }
+		) => {
+			if ( ! isNonNegativeInteger( siteId ) ) {
```
Comment (jblz) [inline client/state/posts/revisions/reducer.js]: siteId actually has to be positive

File: client/state/posts/revisions/schemas.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/** @format */
```
Comment (jblz) [inline client/state/posts/revisions/schemas.js]: This needs updating

File: client/state/posts/revisions/test/reducer.js [hunk 3]
```diff
@@ -9,8 +9,9 @@ import deepFreeze from 'deep-freeze';
 /**
  * Internal dependencies
  */
-import reducer, { requesting, revisions, selection, ui } from '../reducer';
+import reducer, { diffs, requesting, selection, ui } from '../reducer';
```
Comment (jblz) [inline client/state/posts/revisions/test/reducer.js]: reducer tests need updating

File: client/post-editor/editor-revisions-list/index.jsx [hunk 4]
```diff
@@ -111,6 +112,12 @@ class EditorRevisionsList extends PureComponent {
 			'is-loading': isEmpty( revisions ),
 		} );
 
+		// @TODO build these out of the provided `diff` prop
+		const revisionChanges = {
```
Comment (jblz) [inline client/post-editor/editor-revisions-list/index.jsx]: It'd be cool to have the server count these as it goes along as well.

File: client/post-editor/editor-revisions-list/item.jsx [hunk 5]
```diff
@@ -24,10 +24,9 @@ class EditorRevisionsListItem extends PureComponent {
 	};
 
 	render() {
-		const { revision, revisionChanges, isMultiUserSite, translate } = this.props;
-		const authorName = get( revision, 'author.display_name' );
-		const added = get( revisionChanges, 'summary.added' );
-		const removed = get( revisionChanges, 'summary.removed' );
+		const { authorName, revision, revisionChanges, isMultiUserSite, translate } = this.props;
+		const added = get( revisionChanges, 'add', 0 );
```
[the remark this answers is not in the document]
Comment (jblz) [reply client/post-editor/editor-revisions-list/item.jsx]: @spen: the `0` is the `defaultValue` for `get` to cast `undefined` to an int.

**Comment by jblz** [comment]
Rebased for the node upgrade

File: client/post-editor/editor-revisions-list/item.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (spen) [reply client/post-editor/editor-revisions-list/item.jsx]: Ahh, ofc! I missed that it was the default value and was looking at it as if it were part of the path! Thanks for clearing that up :)

**Pull request merged**